src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java

```diff
@@ -200,6 +236,48 @@ public void processElement(final ProcessContext context) throws Exception {
                     });
         }
 
+        /**
+         * Gets the physical location of a given read from the OpticalDuplicateFinder. Caches results for future use
+         * @param read The read to locate.
+         * @return The ReadEndPhysicalLocation for the given read.
+         */
+        private ReadEndPhysicalLocation getPhysicalLocation(Read read,
+            Map<String, ReadEndPhysicalLocation> readNamesToPhysicalLocations) {
+          String readName = read.getFragmentName();
+          if (readNamesToPhysicalLocations.containsKey(readName)) {
+            return readNamesToPhysicalLocations.get(readName);
+          }
+
+          ReadEndPhysicalLocation loc = new ReadEndPhysicalLocation();
+          if (opticalDuplicateFinder.addLocationInformation(readName, loc)) {
+              readNamesToPhysicalLocations.put(readName, loc);
+          }
+          // Should we throw an error if  addLocationInformation fails?
+          return loc;
+        }
+
+        /**
+         * Given PairedEnds and a list of pairs that are duplicates of this pair, determine if the given pair
+         * is an optical duplicate.
+         * @param pair
+         * @param duplicatePairs
+         * @return
+         */
+        private boolean isOpticalDuplicate(final PairedEnds pair, Iterable<PairedEnds> duplicatePairs,
+            Map<String, ReadEndPhysicalLocation> readNamesToPhysicalLocations) {
+
+          ReadEndPhysicalLocation first = getPhysicalLocation(pair.first(), readNamesToPhysicalLocations);
+          ReadEndPhysicalLocation second = getPhysicalLocation(pair.second(), readNamesToPhysicalLocations);
+          for (final PairedEnds duplicatePair : duplicatePairs) {
+            if (opticalDuplicateFinder.isOpticalDuplicate(
+                getPhysicalLocation(duplicatePair.first, readNamesToPhysicalLocations), first) &&
+                opticalDuplicateFinder.isOpticalDuplicate(
```

Comment:
`opticalDuplicateFinder.isOpticalDuplicate` is a bit annoying to read. Fix the naming to make it less verbose